File: gaze/shared/src/test/scala/dev/ligature/gaze/GazeSuite.scala
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */

package dev.ligature.gaze

import munit.Assertions.assertEquals
import munit.FunSuite

class GazeSuite extends FunSuite {
  test("empty input") {
    val gaze = Gaze.from("")
    assert(gaze.isComplete)
    assertEquals(gaze.peek(), Result.NoMatch)
    assertEquals(gaze.peek(), Result.NoMatch)
    assert(gaze.isComplete)
  }

  test("init Gaze with zero values") {
    val gaze = new Gaze(Vector())
    assert(gaze.isComplete)
    assertEquals(gaze.peek(), Result.NoMatch)
    assertEquals(gaze.next(), Result.NoMatch)
    assert(gaze.isComplete)
  }

  test("init Gaze with one value") {
    val gaze = new Gaze(Vector(5));
    assert(!gaze.isComplete)
    assertEquals(gaze.peek(), Result.Match(5))
    assertEquals(gaze.next(), Result.Match(5))
    assert(gaze.isComplete)
    assertEquals(gaze.peek(), Result.NoMatch)
    assertEquals(gaze.next(), Result.NoMatch)
    assert(gaze.isComplete)
  }

  test("init Gaze with single char string") {
    val gaze = Gaze.from("5");
    assert(!gaze.isComplete)
    assertEquals(gaze.peek(), Result.Match('5'))
    assertEquals(gaze.next(), Result.Match('5'))
    assert(gaze.isComplete)
    assertEquals(gaze.peek(), Result.NoMatch)
    assertEquals(gaze.next(), Result.NoMatch)
    assert(gaze.isComplete)
  }
}

File: gaze/src/commonTest/kotlin/dev/ligature/gaze/GazeSuite.kt
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */

package dev.ligature.gaze

import io.kotest.core.spec.style.FunSpec
import io.kotest.matchers.shouldBe

class GazeSuite: FunSpec() {
  init {
    test("empty input") {
      val gaze = Gaze.from("")
      gaze.isComplete shouldBe true
      gaze.peek() shouldBe null
      gaze.peek() shouldBe null
      gaze.isComplete shouldBe true
    }

    test("init Gaze with zero values") {
      val gaze = Gaze(listOf<Unit>())
      gaze.isComplete shouldBe true
      gaze.peek() shouldBe null
      gaze.next() shouldBe null
      gaze.isComplete shouldBe true
    }

    test("init Gaze with one value") {
      val gaze = Gaze(listOf(5))
      gaze.isComplete shouldBe false
      gaze.peek() shouldBe 5
      gaze.next() shouldBe 5
      gaze.isComplete shouldBe true
      gaze.peek() shouldBe null
      gaze.next() shouldBe null
      gaze.isComplete shouldBe true
    }

    test("init Gaze with single char string") {
      val gaze = Gaze.from("5")
      gaze.isComplete shouldBe false
      gaze.peek() shouldBe '5'
      gaze.next() shouldBe '5'
      gaze.isComplete shouldBe true
      gaze.peek() shouldBe null
      gaze.next() shouldBe null
      gaze.isComplete shouldBe true
    }
  }
}
